extensions/amp-script/0.1/user-activation-tracker.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+export const ACTIVATION_TIMEOUT = 5000; // 5 seconds.
+
+const ACTIVATION_EVENTS = ['click', 'input', 'dblclick', 'keypress', 'submit'];
+
+
+/**
+ * See https://github.com/dtapuska/useractivation for inspiration.
+ * @implements {../service.Disposable}
+ */
+export class UserActivationTracker {
+
+  /**
+   * @param {!Element} root
+   */
+  constructor(root) {
+    /** @private @const */
+    this.root_ = root;
+    /** @private @const */
+    this.boundActivated_ = this.activated_.bind(this);
+    /** @private {number} */
+    this.lastActivationTime_ = 0;
+
+    ACTIVATION_EVENTS.forEach(type => {
+      this.root_.addEventListener(
+          type,
+          this.boundActivated_,
+          /* capture */ true);
+    });
+  }
+
+  /** @override */
+  dispose() {
+    ACTIVATION_EVENTS.forEach(type => {
+      this.root_.removeEventListener(
+          type,
+          this.boundActivated_,
+          /* capture */ true);
+    });
+  }
+
+  /**
+   * Whether the element has ever been active since this tracker was alive.
+   * @return {boolean}
+   */
+  hasBeenActive() {
+    return this.lastActivationTime_ > 0;
+  }
+
+  /**
+   * Whether the element is currently considered to be active.
+   * @return {boolean}
+   */
+  isActive() {
+    return this.lastActivationTime_ > 0
```

Comment:
Per spec, this should be reset as soon as it is consumed by something that requires user activation. (and since `amp-script` always requires user-activation to do a run) this pretty much has to reset as soon as the next message is received from worker-dom.